browser/ui/ai_rewriter/ai_rewriter_dialog_delegate.cc

```diff
@@ -0,0 +1,189 @@
+#include "content/public/browser/render_frame_host.h"
+#include "content/public/browser/render_process_host.h"
+#include "content/public/browser/render_widget_host_view.h"
+#include "content/public/browser/weak_document_ptr.h"
+#include "content/public/browser/web_contents.h"
+#include "content/public/browser/web_contents_observer.h"
+#include "content/public/browser/web_ui.h"
+#include "mojo/public/cpp/bindings/remote.h"
+#include "services/service_manager/public/cpp/interface_provider.h"
+#include "ui/aura/window_delegate.h"
+#include "ui/base/ui_base_types.h"
+#include "ui/gfx/geometry/point_f.h"
+#include "ui/gfx/geometry/rect.h"
+#include "ui/gfx/geometry/rect_f.h"
+#include "ui/gfx/geometry/size.h"
+#include "ui/views/widget/widget.h"
+#include "ui/views/widget/widget_delegate.h"
+#include "ui/web_dialogs/web_dialog_delegate.h"
+
+using content::WebContents;
+
+namespace ai_rewriter {
+
+namespace {
+
+gfx::RectF TransformFrameRectToView(content::RenderFrameHost* host,
+                                    const gfx::RectF& bounding_box) {
+  auto* view = host->GetView();
+  if (!view) {
+    return bounding_box;
+  }
+
+  gfx::PointF orig_point(bounding_box.x(), bounding_box.y());
+
+  gfx::PointF transformed_point =
+      view->TransformPointToRootCoordSpaceF(orig_point);
+  return gfx::RectF(transformed_point.x(), transformed_point.y(),
+                    bounding_box.width(), bounding_box.height());
+}
+
+}  // namespace
+
+class AIRewriterDialogDelegate::DialogContentsObserver
+    : public content::WebContentsObserver {
+ public:
+  DialogContentsObserver(content::WebContents* contents,
+                         AIRewriterDialogDelegate* dialog)
+      : content::WebContentsObserver(contents), dialog_(dialog) {}
+  DialogContentsObserver(const DialogContentsObserver&) = delete;
+  DialogContentsObserver& operator=(const DialogContentsObserver&) = delete;
+  ~DialogContentsObserver() override = default;
+
+ private:
+  void WebContentsDestroyed() override { dialog_->ResetDialogObserver(); }
+
+  void PrimaryMainFrameRenderProcessGone(
+      base::TerminationStatus status) override {
+    dialog_->CloseDialog();
+  }
+
+  const raw_ptr<AIRewriterDialogDelegate> dialog_;
+};
+
+// static
+AIRewriterDialogDelegate* AIRewriterDialogDelegate::Show(
+    WebContents* contents) {
+  DCHECK(features::IsAIRewriterEnabled());
+  AIRewriterDialogDelegate* dialog = new AIRewriterDialogDelegate(contents);
+  dialog->ShowDialog();
+  return dialog;
+}
+
+AIRewriterDialogDelegate::AIRewriterDialogDelegate(
+    content::WebContents* contents)
+    : content::WebContentsObserver(contents),
+      target_contents_(contents->GetWeakPtr()) {
+  set_can_close(true);
+  set_dialog_modal_type(ui::MODAL_TYPE_WINDOW);
+  set_dialog_content_url(GURL(kRewriterUIURL));
+  set_dialog_size(gfx::Size(600, 550));
+  set_dialog_args(*base::WriteJson(base::Value::Dict()));
+  set_show_dialog_title(false);
+  set_delete_on_close(false);
```

Comment:
It is my understanding that when we `set_delete_on_close` to false, we should manage it's lifetime by holding it with unique_ptr or something else. I'm curious if this object is deleted properly

Reply:
![image](https://github.com/brave/brave-core/assets/7678024/97703c17-05bb-4a5b-9fae-f62302525ecc)

I added a log to the constructor to make sure and this is definitely destroyed when the dialog is closed. This is moved into |ShowConstrainedWebDialog| via `WrapUnique` - I copied the pattern from `chrome/browser/ui/webui/signin/signin_email_confirmation_dialog.cc`

If I don't set `set_delete_on_close(false)` I get a crash where we attempt to free this twice.

Reply:
Great! Thanks for checking